src/Settings/Builders/RunTest.hs

```diff
@@ -81,53 +72,76 @@ runTestBuilderArgs = builder RunTest ? do
+                           Nothing -> []
+        onlyPerfArg  = if testOnlyPerf args
+                           then Just "--only-perf-tests"
+                           else Nothing
+        skipPerfArg  = if testSkipPerf args
+                           then Just "--skip-perf-tests"
+                           else Nothing
+        speedArg     = ["-e", "config.speed=" ++ setTestSpeed (testSpeed args)]
+        summaryArg   = case testSummary args of
+                           Just filepath -> Just $ "--summary-file" ++ quote filepath
+                           Nothing -> Just $ "--summary-file=testsuite_summary.txt"
+        junitArg     = case testJUnit args of
+                           Just filepath -> Just $ "--junit " ++ quote filepath
+                           Nothing -> Nothing
+        configArgs   = concat [["-e", configArg] | configArg <- testConfigs args]
         verbosityArg = case testVerbosity args of
-                         Nothing -> Nothing
-                         Just verbosity -> Just $ "--verbose=" ++ verbosity
-        wayArgs    = map ("--way=" ++) (testWays args) 
-    pure $  testOnlyArg
-         ++ speedArg 
+                           Nothing -> Just $ "--verbose=" ++ show (fromEnum globalVerbosity)
+                           Just verbosity -> Just $ "--verbose=" ++ verbosity
+        wayArgs      = map ("--way=" ++) (testWays args) 
+        compilerArg  = ["--config", "compiler=" ++ show (compiler)]
+        ghcPkgArg    = ["--config", "ghc-pkg=" ++ show (bindir -/- "ghc-pkg")]
+        haddockArg   = ["--config", "haddock=" ++ show (bindir -/- "haddock")]
+        hp2psArg     = ["--config", "hp2ps=" ++ show (bindir -/- "hp2ps")]
+        hpcArg       = ["--config", "hpc=" ++ show (bindir -/- "hpc")]   
+    pure $  testOnlyArg ++ speedArg 
          ++ catMaybes [ onlyPerfArg, skipPerfArg, summaryArg
                       , junitArg, verbosityArg  ] 
-         ++ configArgs
-         ++ wayArgs
+         ++ configArgs ++ wayArgs ++  compilerArg ++ ghcPkgArg
+         ++ haddockArg ++ hp2psArg ++ hpcArg
+
+-- | Directory to look for Binaries
+setBinaryDirectory :: String -> Action FilePath
+setBinaryDirectory "stage0" = setting InstallBinDir
+setBinaryDirectory "stage1" = liftM2 (-/-) topDirectory (stageBinPath Stage0) 
+setBinaryDirectory "stage2" = liftM2 (-/-) topDirectory (stageBinPath Stage1) 
+setBinaryDirectory compiler = pure $ parentPath compiler
+
+-- | Set Test Compiler
+setCompiler :: String -> Action FilePath
+setCompiler "stage0" = setting SystemGhc
+setCompiler "stage1" = liftM2 (-/-) topDirectory (fullpath Stage0 ghc)
+setCompiler "stage2" = liftM2 (-/-) topDirectory (fullpath Stage1 ghc)
+setCompiler compiler = pure compiler 
 
 -- | Set speed for test
 setTestSpeed :: TestSpeed -> String
 setTestSpeed Fast    = "2"
 setTestSpeed Average = "1"
 setTestSpeed Slow    = "0"
 
+-- | Returns parent path of test compiler 
+parentPath :: String -> String
```

Comment:
Such questions should also go in the source code in the form of `TODO`s. For example:
```haskell
-- TODO: The current implementation is not perfect, because X, Y and Z. Can we find a better one?
```